### PR TITLE
Support fan status monitor, temperature monitor and thermal policy control

### PR DESCRIPTION
**Pull request by Junchao-Mellanox**
What I did?
1. Add FanUpdater to get fan information and write it to database
2. Add TemperUpdater to get temperature information and write it to database
3. Add ThermalControlDaemon to run the thermal policy 

How to verify?
1. show platform fanstatus
2. show platform temperature
3. View syslog make sure thermalctld is up and there is no error logs

### CODE REVIEW

File: sonic-thermalctld/scripts/thermalctld [hunk 1]
```diff
@@ -0,0 +1,521 @@
...
+"""
+
+try:
+    import os
+    import sys
+    import time
+    import signal
+    import threading
+    from datetime import datetime
+    #from swsscommon import swsscommon
```
Comment (Junchao-Mellanox) [inline sonic-thermalctld/scripts/thermalctld]: Remove this line.

File: sonic-thermalctld/scripts/thermalctld [hunk 2]
```diff
@@ -0,0 +1,521 @@
...
+        """
+        Constructor for FanUpdater
+        :param chassis: Object representing a platform chassis
+        """
+        self.chassis = chassis
+        self.fan_status_dict = {}
+        state_db = daemon_base.db_connect(swsscommon.STATE_DB)
+        self.table = swsscommon.Table(state_db, FanUpdater.FAN_INFO_TABLE_NAME)
+
+    def destroy(self):
```
Comment (Junchao-Mellanox) [inline sonic-thermalctld/scripts/thermalctld]: deinit

File: sonic-thermalctld/scripts/thermalctld [hunk 3]
```diff
@@ -0,0 +1,521 @@
...
+                # TODO: wait for Kebo to define the mapping of fan status to led color,
+                # just set it to red so far
+                fan.set_status_led(fan.STATUS_LED_COLOR_RED)
+        except NotImplementedError as e:
+            logger.log_warning('Failed to set led to fan, set_status_led not implemented')
+
+
+class TemperStatus(object):
+    def __init__(self):
+        self.over_temper = False
```
Comment (Junchao-Mellanox) [inline sonic-thermalctld/scripts/thermalctld]: temper ->temperature

File: sonic-thermalctld/scripts/thermalctld [hunk 4]
```diff
@@ -0,0 +1,521 @@
...
+        """
+        Constructor of TemperUpdater
+        :param chassis: Object representing a platform chassis
+        """
+        self.chassis = chassis
+        self.temper_status_dict = {}
+        state_db = daemon_base.db_connect(swsscommon.STATE_DB)
+        self.table = swsscommon.Table(state_db, TemperUpdater.TEMPER_INFO_TABLE_NAME)
+
+    def destroy(self):
```
Comment (Junchao-Mellanox) [inline sonic-thermalctld/scripts/thermalctld]: deinit

File: sonic-thermalctld/scripts/thermalctld [hunk 5]
```diff
@@ -0,0 +1,521 @@
...
+        self.temper_updater.destroy()
+
+        logger.log_info("Stop thermal monitoring loop")
+
+
+#
+# Daemon =======================================================================
+#
+class ThermalControlDaemon(DaemonBase):
+    # Interval to run thermal control logical
```
Comment (Junchao-Mellanox) [inline sonic-thermalctld/scripts/thermalctld]: logical->logic

File: sonic-thermalctld/scripts/thermalctld [hunk 6]
```diff
@@ -0,0 +1,521 @@
...
+            logger.log_warning("Caught unhandled signal '" + sig + "'")
+
+    def run(self):
+        """
+        Run main logical of this daemon
+        :return:
+        """
+        logger.log_info("Starting up...")
+
+        import sonic_platform.platform
```
Comment (Junchao-Mellanox) [inline sonic-thermalctld/scripts/thermalctld:527]: try move it to top

File: sonic-thermalctld/scripts/thermalctld [hunk 7]
```diff
@@ -0,0 +1,521 @@
...
+
+    # Signal handler
+    def signal_handler(self, sig, frame):
+        """
+        Signal handler
+        :param sig: Signal number
+        :param frame: not used
+        :return:
+        """
+        if sig == signal.SIGHUP:
```
Comment (keboliu) [inline sonic-thermalctld/scripts/thermalctld:509]: please test and make sure thread will also exit on SIGTERM

File: sonic-thermalctld/scripts/thermalctld [hunk 8]
```diff
@@ -0,0 +1,521 @@
...
+        if presence == self.presence:
+            return False
+
+        self.presence = presence
+        return True
+
+    def _check_speed_value_available(self, speed, target_speed, tolerance, current_status):
+        if speed == NOT_AVAILABLE or target_speed == NOT_AVAILABLE or tolerance == NOT_AVAILABLE:
+            if current_status is True:
+                logger.log_warning('Fan speed or target_speed or tolerance become unavailable.')
```
Comment (keboliu) [inline sonic-thermalctld/scripts/thermalctld]: make log more specific to indicate which speed is not available.

File: sonic-thermalctld/scripts/thermalctld [hunk 9]
```diff
@@ -0,0 +1,521 @@
...
+        :param target_speed: Fan target speed
+        :param tolerance: Threshold between Fan speed and target speed
+        :return: True if status changed else False
+        """
+        if not self._check_speed_value_available(speed, target_speed, tolerance, self.under_speed):
+            old_status = self.under_speed
+            self.under_speed = False
+            return old_status != self.under_speed
+
+        status = (target_speed - speed > tolerance)
```
Comment (keboliu) [inline sonic-thermalctld/scripts/thermalctld]: this is not the correct way to check over speed or under-speed, should be:    target_speed*(1-tolerance) < speed < target_speed*(1+tolerance)

File: sonic-thermalctld/scripts/thermalctld [hunk 10]
```diff
@@ -0,0 +1,521 @@
...
+        :param target_speed: Fan target speed
+        :param tolerance: Threshold between Fan speed and target speed
+        :return: True if status changed else False
+        """
+        if not self._check_speed_value_available(speed, target_speed, tolerance, self.over_speed):
+            old_status = self.over_speed
+            self.over_speed = False
+            return old_status != self.over_speed
+
+        status = (speed - target_speed > tolerance)
```
Comment (keboliu) [inline sonic-thermalctld/scripts/thermalctld]: as above comments

**Review by keboliu — CHANGES_REQUESTED** [review]
as comments

**Pull request closed without merging**